File: kmp-native/shared/src/commonMain/kotlin/com/stackbenchmark/kmpnative/data/mapper/Mappers.kt
package com.stackbenchmark.kmpnative.data.mapper

import com.stackbenchmark.kmpnative.data.dto.GenreDto
import com.stackbenchmark.kmpnative.data.dto.MovieDetailDto
import com.stackbenchmark.kmpnative.data.dto.MovieDto
import com.stackbenchmark.kmpnative.data.dto.PopularMoviesResponseDto
import com.stackbenchmark.kmpnative.domain.entity.Genre
import com.stackbenchmark.kmpnative.domain.entity.Movie
import com.stackbenchmark.kmpnative.domain.entity.MovieDetail
import com.stackbenchmark.kmpnative.domain.entity.MoviePage

fun MovieDto.toDomain(): Movie = Movie(
    id = id,
    title = title,
    overview = overview,
    posterPath = posterPath,
    backdropPath = backdropPath,
    releaseDate = releaseDate,
    voteAverage = voteAverage,
    voteCount = voteCount
)

fun MovieDetailDto.toDomain(): MovieDetail = MovieDetail(
    id = id,
    title = title,
    overview = overview,
    posterPath = posterPath,
    backdropPath = backdropPath,
    releaseDate = releaseDate,
    voteAverage = voteAverage,
    voteCount = voteCount,
    tagline = tagline,
    runtime = runtime,
    genres = genres.map { it.toDomain() }
)

fun GenreDto.toDomain(): Genre = Genre(
    id = id,
    name = name
)

fun PopularMoviesResponseDto.toDomain(): MoviePage = MoviePage(
    movies = results.map { it.toDomain() },
    page = page,
    totalPages = totalPages
)

File: rn-expo/src/data/mappers.ts
import { Genre, Movie, MovieDetail, MoviePage } from "../domain/entities";
import {
  GenreDto,
  MovieDetailDto,
  MovieDto,
  PopularMoviesResponseDto,
} from "./dtos";

export function mapMovieDtoToDomain(dto: MovieDto): Movie {
  return {
    id: dto.id,
    title: dto.title,
    overview: dto.overview,
    posterPath: dto.poster_path,
    backdropPath: dto.backdrop_path,
    releaseDate: dto.release_date,
    voteAverage: dto.vote_average,
    voteCount: dto.vote_count,
  };
}

export function mapGenreDtoToDomain(dto: GenreDto): Genre {
  return {
    id: dto.id,
    name: dto.name,
  };
}

export function mapMovieDetailDtoToDomain(dto: MovieDetailDto): MovieDetail {
  return {
    ...mapMovieDtoToDomain(dto),
    tagline: dto.tagline,
    runtime: dto.runtime,
    genres: dto.genres.map(mapGenreDtoToDomain),
  };
}

export function mapPopularMoviesResponseToDomain(
  dto: PopularMoviesResponseDto
): MoviePage {
  return {
    movies: dto.results.map(mapMovieDtoToDomain),
    page: dto.page,
    totalPages: dto.total_pages,
  };
}
